drivers/flash/flash_sam0.c

```diff
@@ -238,23 +240,34 @@ static int flash_sam0_write(const struct device *dev, off_t offset,
 
 	flash_sam0_sem_take(dev);
 
-	for (addr = offset; addr < offset + len; addr++) {
-		off_t base = addr & ~(ROW_SIZE - 1);
+	err = flash_sam0_write_protection(dev, false);
+	if (err == 0) {
+		for (addr = offset; addr < offset + len; addr++) {
+			off_t base = addr & ~(ROW_SIZE - 1);
+
+			if (base != ctx->offset) {
+				/* Started a new row. Flush any pending ones. */
+				flash_sam0_commit(dev);
+				memcpy(ctx->buf, (void *)base,
+				       sizeof(ctx->buf));
+				ctx->offset = base;
+			}
 
-		if (base != ctx->offset) {
-			/* Started a new row.  Flush any pending ones. */
-			flash_sam0_commit(dev);
-			memcpy(ctx->buf, (void *)base, sizeof(ctx->buf));
-			ctx->offset = base;
+			ctx->buf[addr % ROW_SIZE] = *pdata++;
 		}
 
-		ctx->buf[addr % ROW_SIZE] = *pdata++;
+		flash_sam0_commit(dev);
+	}
+
+	int err2 = flash_sam0_write_protection(dev, true);
```

Comment:
Shouldn't this be also inside the `if (err == 0)` block? If `err != 0` we have failed to unlock the flash anyway, the lock will probably also fail.

Reply:
My point is to not leave memory as unprotected if possible. It might be (quite probably) that it not a case after unprotect call failed, but calling the protect shouldn't make any additional harm.